File: aw_scraper-main/ackers_weldon/pipelines/sharesight/news.py
from itemadapter import ItemAdapter
import pymongo
import datetime

class NewsPipeline:
    collection_name = "news"

    def __init__(self, mongo_uri, mongo_db):
        self.mongo_uri = mongo_uri
        self.mongo_db = mongo_db

    @classmethod
    def from_crawler(cls, crawler):
        return cls(
            mongo_uri=crawler.settings.get("MONGO_URI"),
            mongo_db=crawler.settings.get("MONGO_DATABASE", "items"),
        )

    def open_spider(self, spider):
        self.client = pymongo.MongoClient(self.mongo_uri)
        self.db = self.client[self.mongo_db]

    def close_spider(self, spider):
        self.client.close()

    def process_item(self, item, spider):
        domain = "https://www.sharesight.com/blog/"
                
        try:

            title = item.get("title", {}).get("title") or ""
            summary = item.get("metaDescription", {}).get("metaDescription") or ""

            
            categories = item.get("categories") or []

            news = {
                "_id": domain + item.get("urlSlug", ""),
                "title": title,
                "summary": summary,
                "date":  item.get("created_at", datetime.datetime.now().isoformat()),
                "source": "sharesight",
                "categories": list(set([entry.get("name", {}).get("name", "") for entry in categories])),
                "type": "blog",
                "image": None,
                "read": False
            }
            try:
                news["date"] = datetime.datetime.strptime(news["date"], "%Y-%m-%dT%H:%M%z").timestamp()
            except:
                news["date"] = datetime.datetime.strptime(news["date"], "%Y-%m-%d").timestamp()
            images = item.get("image", {}).get("gatsbyImageData", {}).get("images", {})
            news["image"] = images.get("sources", []).pop().get("src") or images.get("fallback", {}).get("src")
            news["image"] = news["image"].replace(" ", "%20")

            self.db[self.collection_name].insert_one(ItemAdapter(news).asdict())
        except pymongo.errors.DuplicateKeyError:
            spider.crawler.engine.close_spider(self, reason='cancelled')

        return news

File: dashboard-aw-main/models/news/News.ts
import { models, model, Schema } from 'mongoose';

export interface INews extends Document {
    _id: string;
    title: string;
    summary?: string;
    date: number;
    source: string;
    categories: string[];
    image?: string;
    type: string;
    ai_summary?: string;
    sentiment?: number;
    read?: boolean;
}

const newsSchema: Schema = new Schema({
    _id: { type: String, required: true },
    title: { type: String, required: true },
    summary: { type: String, required: false },
    date: { type: Number, required: true },
    source: { type: String, required: true },
    categories: { type: Array, required: false },
    image: { type: String, required: false },
    type: { type: String, required: true },
    ai_summary: { type: String, required: false },
    sentiment: { type: Number, required: false },
    read: { type: Boolean, required: false, default: false }
});

const News = models.News || model<INews>('News', newsSchema);
export default News;